File: project/jimmer-apt/src/main/java/org/babyfish/jimmer/apt/immutable/ImmutableProcessor.java
package org.babyfish.jimmer.apt.immutable;

import site.addzero.context.Context;
import site.addzero.context.LsiSourceFilterKt;
import site.addzero.context.Settings;
import site.addzero.lsi.clazz.LsiClass;
import site.addzero.lsi.codegen.GeneratedResourceArtifact;
import site.addzero.lsi.jimmer.immutable.metadata.extractor.ImmutableCollectedSourceAccumulator;
import site.addzero.lsi.jimmer.immutable.metadata.extractor.ImmutableCollectedSourceResolution;
import site.addzero.lsi.jimmer.immutable.metadata.generator.ImmutableGeneratedOutput;
import site.addzero.lsi.jimmer.immutable.metadata.generator.ImmutableProcessorSupport;
import site.addzero.lsi.poet.LsiFileSpec;

import java.io.File;
import java.util.ArrayList;
import java.util.Collection;
import java.util.function.Consumer;

public class ImmutableProcessor {

    private final boolean buddyIgnoreResourceGeneration;

    public ImmutableProcessor(boolean buddyIgnoreResourceGeneration) {
        this.buddyIgnoreResourceGeneration = buddyIgnoreResourceGeneration;
    }

    public Collection<LsiClass> process() {
        ImmutableCollectedSourceResolution resolvedSources = resolveImmutableSources();
        generateJimmerTypes(resolvedSources);
        return new ArrayList<>(resolvedSources.getLsiClasses());
    }

    private ImmutableCollectedSourceResolution resolveImmutableSources() {
        ImmutableCollectedSourceAccumulator accumulator = new ImmutableCollectedSourceAccumulator();
        ImmutableProcessorSupport.collectRoundSources(
                accumulator,
                Context.INSTANCE.getLsiResolver(),
                LsiSourceFilterKt::matchesConfiguredSourceFilters
        );
        ImmutableCollectedSourceResolution resolvedSources = ImmutableProcessorSupport.resolveCollectedSources(
                accumulator,
                Context.INSTANCE.getLsiResolver(),
                Context.INSTANCE::typeOf
        );
        if (ImmutableProcessorSupport.hasImmutableTypes(resolvedSources)) {
            Context.INSTANCE.resolve();
        }
        return resolvedSources;
    }

    private void generateJimmerTypes(ImmutableCollectedSourceResolution resolvedSources) {
        if (!ImmutableProcessorSupport.hasImmutableTypes(resolvedSources)) {
            return;
        }
        ImmutableGeneratedOutput generatedOutput = ImmutableProcessorSupport.generateAptOutput(
                resolvedSources,
                Settings.INSTANCE.getJimmerExcludedUserAnnotationPrefixes(),
                Context.INSTANCE.getJacksonTypes(),
                findGeneratedEntitiesResourceFile(),
                buddyIgnoreResourceGeneration,
                Settings.INSTANCE.getJimmerImmutableIsModuleRequired(),
                org.babyfish.jimmer.JimmerVersionsKt.currentVersion()
        );
        for (LsiFileSpec fileSpec : generatedOutput.getSourceFileSpecs()) {
            writeFileSpec(fileSpec);
        }
        for (GeneratedResourceArtifact artifact : generatedOutput.getResourceArtifacts()) {
            writeResourceArtifact(artifact);
        }
        ImmutableProcessorSupport.logResolvedImmutableTypes(
                resolvedSources.getLsiClasses(),
                new Consumer<String>() {
                    @Override
                    public void accept(String message) {
                        Context.INSTANCE.logInfo(message);
                    }
                }
        );
        ImmutableProcessorSupport.notifyEntityMetaConsumers(
                resolvedSources.getLsiClasses(),
                new Consumer<String>() {
                    @Override
                    public void accept(String message) {
                        Context.INSTANCE.logInfo(message);
                    }
                }
        );
    }

    private File findGeneratedEntitiesResourceFile() {
        return Context.INSTANCE.guessGeneratedJimmerResourceFile("entities");
    }

    private void writeFileSpec(LsiFileSpec fileSpec) {
        Context.INSTANCE.getLsiFiler().createSourceFile(fileSpec);
    }

    private void writeResourceArtifact(GeneratedResourceArtifact resourceArtifact) {
        Context.INSTANCE.getLsiFiler().createResourceFile(resourceArtifact.getPath(), resourceArtifact.getContent());
    }
}

File: project/compiler/immutable/jimmer-ksp-immutable/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/ImmutableProcessor.kt
package org.babyfish.jimmer.ksp.immutable

import com.google.auto.service.AutoService
import site.addzero.context.Context
import site.addzero.lsi.codegen.GeneratedResourceArtifact
import site.addzero.lsi.jimmer.immutable.metadata.extractor.ImmutableCollectedSourceAccumulator
import site.addzero.lsi.jimmer.immutable.metadata.generator.ImmutableProcessorSupport
import site.addzero.context.Settings
import site.addzero.context.matchesConfiguredSourceFilters
import site.addzero.lsi.poet.LsiFileSpec
import site.addzero.lsi.processor.ProcessorSpi

@AutoService(ProcessorSpi::class)
class ImmutableProcessor : ProcessorSpi<Context, Unit> {
    override var ctx = Context
    private val collectedSources = ImmutableCollectedSourceAccumulator()

    override fun onRound() {
        // 覆盖来源：project/compiler/immutable/jimmer-ksp-immutable/.../ImmutableProcessor.process
        // 原 process() 单阶段“扫描+生成”路径中的扫描与校验部分，迁移到 onRound() 多轮收集阶段
        ImmutableProcessorSupport.collectRoundSources(
            accumulator = collectedSources,
            resolver = ctx.lsiResolver,
            include = java.util.function.Predicate { it.matchesConfiguredSourceFilters() },
        )
    }

    override fun onFinish() {
        // 覆盖来源：project/compiler/immutable/jimmer-ksp-immutable/.../ImmutableProcessor.process
        // 原 process() 单阶段“扫描+生成”路径中的代码生成与 SPI 通知，迁移到 onFinish() 收尾阶段统一执行
        if (collectedSources.isEmpty()) {
            return
        }
        val resolvedSources = ImmutableProcessorSupport.resolveCollectedSources(
            accumulator = collectedSources,
            resolver = ctx.lsiResolver,
            toImmutableType = java.util.function.Function(Context::typeOf),
        )
        if (!ImmutableProcessorSupport.hasImmutableTypes(resolvedSources)) {
            // 覆盖来源：project/compiler/immutable/jimmer-ksp-immutable/.../ImmutableProcessor.onFinish 的 `immutableTypeMultiMap.isEmpty()` 早退
            // 迁移说明：旧的并行 `Map<String, List<ImmutableType>>` 判空逻辑改为基于 resolved-source 聚合结构判断，保持“无可生成 immutable 类型时直接结束”的行为
            clearCollected()
            return
        }
        ctx.resolve()
        val generatedOutput = ImmutableProcessorSupport.generateKspOutput(
            resolvedSources = resolvedSources,
            // 覆盖来源：project/compiler/immutable/jimmer-ksp-immutable/.../ImmutableProcessor.generateJimmerTypes
            // 迁移说明：processor finish-stage 只消费一次解析后的 collected-source resolution，再统一落盘 artifact
            excludedUserTypePrefixes = Settings.jimmerExcludedUserAnnotationPrefixes,
            jacksonTypes = ctx.jacksonTypes,
            existingEntitiesResourceFile = ctx.guessGeneratedJimmerResourceFile("entities"),
            isResourceGenerationIgnored = Settings.jimmerBuddyIgnoreResourceGeneration,
            isModuleRequired = Settings.jimmerImmutableIsModuleRequired,
        )
        generatedOutput.sourceFileSpecs.forEach(::writeFileSpec)
        generatedOutput.resourceArtifacts.forEach(::writeResourceArtifact)
        ImmutableProcessorSupport.notifyEntityMetaConsumers(
            entities = resolvedSources.lsiClasses,
            infoLogger = java.util.function.Consumer { message ->
                ctx.logInfo(message)
            },
        )
        clearCollected()
    }

    private fun writeFileSpec(fileSpec: LsiFileSpec) {
        ctx.lsiFiler.createSourceFile(fileSpec)
    }

    private fun writeResourceArtifact(artifact: GeneratedResourceArtifact) {
        // 覆盖来源：project/compiler/immutable/jimmer-ksp-immutable/.../JimmerModuleGenerator.generate entities 资源输出
        // 迁移说明：immutable 顶层资源写入统一收口到 processor，generator 只返回纯 resource artifact
        ctx.lsiFiler.createResourceFile(artifact.path, artifact.content)
    }

    private fun clearCollected() {
        collectedSources.clear()
    }
}
